Use storage backend with attachments

fabrixxm opened the pull request.

fabrixxm committed:
Move mod/attach to src, add Attach model, update attach table to use storage backends

fabrixxm committed:
Attach: store, update, delete. Model and views

fabrixxm committed:
Photo model. Small fix to `update()`

pass `$old_fields` param to `DBA::update()`

fabrixxm committed:
Fix Attach model store() and storeFile()

fabrixxm committed:
Item model, fix deletion of all attachments

Item model delete all attachments on item deletion, but it should check
if same attachment isn't used somewhere else.

fabrixxm committed:
Use Attach model while deleting in video tab

fabrixxm committed:
Fix undefined var in module Proxy

fabrixxm committed:
Use more Attach model

fabrixxm committed:
return default mimetype for unknown extensions

fabrixxm committed:
Update "storage" console command

add support to move data in 'attach' table

fabrixxm committed:
update database.sql

fabrixxm committed:
remove file added by mistake

MrPetovan commented on mod/videos.php:
```diff
@@ -150,25 +116,16 @@ function videos_post(App $a)
-			DBA::escape($video_id)
-		);
+		$r = Attach::exists(['id' => $video_id, 'uid' => local_user()]);
+		if ($r === true) {
```
You don't need the intermediate variable here, you can put the `Attach::exists` call directly in the `if ()`.

MrPetovan commented on src/Model/Attach.php:
```diff
@@ -0,0 +1,289 @@
+	public static function selectFirst(array $fields = [], array $conditions = [], array $params = [])
+	{
+		if (empty($fields)) {
+			$fields = self::getFields();
```
`DBA::select` already has a provision for empty field list which means "all the fields in the queried table". You don't need to extract the `attach` table fields and explicitly pass them, you can directly pass the `$fields` parameter to `DBA::selectFirst`.

fabrixxm replied:
yup, but `self::getFields()` excludes "data" column. Attach model (like Photo model) doesn't return file data while you are looking for metadata. You have to explicitly call `Attach::getData(..)` which, if it's the case, interact with storage backend.

MrPetovan replied:
Oooooh, I missed that part. Very well then.

MrPetovan commented on src/Module/Attach.php:
```diff
@@ -0,0 +1,67 @@
+	 *
+	 * Fetch an attached file given the id
+	 */
+	public static function init()
```
Given the nature of this method outputting content directly and exiting, this should rather be `rawContent`to avoid confusion.

MrPetovan reviewed (changes requested):
This is seriously awesome, thanks for your work! There are a few minor improvements I could see.

annando commented on src/Core/Console/Storage.php:
```diff
@@ -19,9 +19,18 @@ protected function getHelp()
+        Show this help
+    
     bin/console storage list
+        List avaiable storage backends
```
"available"

fabrixxm committed:
Small fixes

- Avoid an intermediate variable
- Use `rawContent()` in `Attach` module
- Small typo
